src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -166,37 +166,50 @@ private BoundPattern BindDiscardPattern(DiscardPatternSyntax node, TypeSymbol op
             return new BoundConstantPattern(node, convertedExpression, constantValueOpt ?? ConstantValue.Bad, hasErrors);
         }
 
-        internal BoundExpression ConvertPatternExpression(TypeSymbol inputType, CSharpSyntaxNode node, BoundExpression expression, ref ConstantValue constantValue, DiagnosticBag diagnostics)
+        internal BoundExpression ConvertPatternExpression(TypeSymbol inputType, CSharpSyntaxNode node, BoundExpression expression, out ConstantValue constantValue, DiagnosticBag diagnostics)
         {
-            // NOTE: This will allow user-defined conversions, even though they're not allowed here.  This is acceptable
-            // because the result of a user-defined conversion does not have a ConstantValue and we'll report a diagnostic
-            // to that effect later.
-            BoundExpression convertedExpression = GenerateConversionForAssignment(inputType, expression, diagnostics);
+            BoundExpression convertedExpression;
 
-            if (convertedExpression.Kind == BoundKind.Conversion)
+            // If we are pattern-matching against an open type, we do not convert the constant to the type of the input.
+            // This permits us to match a value of type `IComparable<T>` with a pattern of type `int`.
+            if (inputType.ContainsTypeParameter() &&
+                // But do not permit matching null against a struct type.
+                (expression.ConstantValue != ConstantValue.Null || !inputType.IsNonNullableValueType()))
```

Comment:
If `inputType` is known at compile-time not to be convertible to the expression type, are we reporting an error?

Reply:
Not if the input type is an open type. We need to be able to match an expression of type `IComparable<T>` to `3` even though there is no conversion between them. That change is the purpose of this PR.

---
In reply to: [180636955](https://github.com/dotnet/roslyn/pull/25995#discussion_r180636955) [](ancestors = 180636955)

Reply:
Ideally we should generate a warning for cases such as `new C<T>() is 3`. Please add a test.

---
In reply to: [180840789](https://github.com/dotnet/roslyn/pull/25995#discussion_r180840789) [](ancestors = 180840789,180636955)

Reply:
Done in a new iteration.

---
In reply to: [180844755](https://github.com/dotnet/roslyn/pull/25995#discussion_r180844755) [](ancestors = 180844755,180840789,180636955)